### PR TITLE
Improve efficiency of BoundedBreakIteratorScanner fragmentation algorithm

### PR DESCRIPTION
**Pull request by thelink2012**
As discussed in #73569 the current implementation is too slow in certain scenarios.

The [inefficient part](https://github.com/elastic/elasticsearch/blob/a92a647b9f17d1bddf5c707490a19482c273eda3/server/src/main/java/org/apache/lucene/search/uhighlight/BoundedBreakIteratorScanner.java#L92-L101) of the code can be stated as the following problem:

> Given a text (`getText()`) and a position in this text (`offset`), find the sentence boundary before and after the offset, in such a way that the after boundary is maximal but respects `end boundary - start boundary < fragment size`.
>
> In case it's impossible to produce an after boundary that respects the said condition, use the nearest boundary following offset.

The current approach begins by finding the nearest preceding and following boundaries, and expands the following boundary greedily while it respects the problem restriction. This is fine asymptotically, but `BreakIterator` which is used to find each boundary is sometimes expensive \[[Lucene noticed too](https://github.com/apache/lucene/blob/d5d6dc079395c47cd6d12dcce3bcfdd2c7d9dc63/lucene/highlighter/src/java/org/apache/lucene/search/uhighlight/LengthGoalBreakIterator.java#L202)\].

The new approach maximizes the after boundary by scanning for the last boundary preceding the position that would cause the condition to be violated (i.e. knowing `start boundary` and `offset`, how many characters are left before resulting length is `fragment size`). If this scan finds the start boundary, it means it's impossible to satisfy the problem restriction, and we get the first boundary following offset instead (or better, since we already scanned `[offset, targetEndOffset]`, start from `targetEndOffset + 1`).

In theory, both approaches should produce exactly the same outputs given the same `(text, offset, fragment_size)` tuple. But it doesn't. As far as my investigation went, `BreakIterator` doesn't seem to be commutative. Previous method calls affects other calls. So I guess whatever change we make to this algorithm, may produce results that differ from each other. It is rare, but happens. Would love to be wrong on this though.

Highlighting related unit and integration tests pass. Though I'm not passing in tests that seem completely unrelated to this. See below.

### CODE REVIEW

**Comment by thelink2012** [comment]
We're still working on the CLA thing. But back to the test issues, [here's](https://gist.github.com/thelink2012/24c5bb08968ccc07da16d06ba14e4340) what I'm getting after running `./gradlew check -Dtests.haltonfailure=false `.

**Review by jimczi — COMMENTED** [review]
I left one comment regarding an edge case but the logic looks good to me.

> But back to the test issues, here's what I'm getting after running ./gradlew check -Dtests.haltonfailure=false . Seems to be related to distribution and documentation. Should I worry about it?

The integration tests for docs seem to fail. It could be a side effect of this change so the best is to look at the report that is linked after the error message:
`Execution failed for task ':docs:integTest'.`

> We're still working on the CLA thing.

Thanks, don't hesitate to ping me on the PR when it's done.

> In theory, both approaches should produce exactly the same outputs given the same (text, offset, fragment_size) tuple. But it doesn't

I am surprised too. I'll dig to see where that comes from.

File: server/src/main/java/org/apache/lucene/search/uhighlight/BoundedBreakIteratorScanner.java
Comment (thelink2012) [reply server/src/main/java/org/apache/lucene/search/uhighlight/BoundedBreakIteratorScanner.java:107]: You are right. The only way for `innerStart == innerEnd` to hold true for this case is when `text` is empty. In this case, the code would break without the condition. But given this is coded specifically for the unified highlighter, that doesn't seem like a possibility. Is that correct?

**Comment by thelink2012** [comment]
> The integration tests for docs seem to fail. It could be a side effect of this change so the best is to look at the report that is linked after the error message:

Happens on master too, before the patches. I rebased to 2c521279e4e0180f7a41d15b39c1fd4bd69032e7 and issue still occurs. Reports can be found [here](https://drive.google.com/file/d/1kNGtRu4FiUi0-oJUAoHPdI-IeER9Og9y/view?usp=sharing). It is related to a timeout in `DocsClientYamlTestSuiteIT`:

```
java.lang.Exception: Suite timeout exceeded (>= 2400000 msec).
	at __randomizedtesting.SeedInfo.seed([7C03B7D75C3D1D4D]:0)
```

Haven't tried increasing the timeout since 40 minutes is a lot already.

**Comment by jimczi** [comment]
Sorry for the slow reply. We don't see these timeouts in our build but I doubt that this change is responsible. In any case we'll trigger all tests in our CI before merging so if there's a problem we will see it.
Any news on the CLA front ?

**Comment by thelink2012** [comment]
No problem. Sorry for the delay on the CLA too. I'll ping you once that's signed :)

**Comment by thelink2012** [comment]
Hey @jimczi. We've signed the CLA. Could you check?

**Comment by jimczi** [comment]
@elasticmachine update branch

**Review by jimczi — COMMENTED** [review]
Thanks @thelink2012 , the change looks good to me.
Can you merge master in your branch so that I can trigger the tests with the latest changes ?

**Comment by thelink2012** [comment]
Done.

**Comment by jimczi** [comment]
@elasticmachine ok to test

**Comment by jimczi** [comment]
@thelink2012 can you merge master again ? Sorry for the back and forth but the last merge was made on a broken state.

**Comment by thelink2012** [comment]
No problem. Merged. Let's see how the tests go:)

**Comment by jimczi** [comment]
@elasticmachine ok to test

**Review by jimczi — APPROVED** [review]
LGTM, thanks @thelink2012 !

**Comment by tylersmalley** [comment]
@jimczi, looks like this change caused issues that were caught in Kibana testing - preventing us from using the nightly snapshots. I think it would be best to revert this PR to unblock downstream teams relying on an up-to-date Elasticsearch with Kibana. If you agree that is the correct approach, would you mind helping with a revert here? I am headed out on vacation right now, but will be back Monday and can assist with validating a new PR against Kibana going forward.

**Comment by ywelsch** [comment]
@jimczi is currently out on vacation. I've reverted the relevant commits on master (154105f48e1) / 7.x (e97c4ce5e8b) branches to unblock downstream teams.

@thelink2012 can you open up another separate PR again for this where we can then iterate on the necessary fixes?

**Comment by thelink2012** [comment]
Done, @ywelsch

**Pull request merged**
